<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>MendelPlots.jl · MendelPlots</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link href="assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>MendelPlots</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li class="current"><a class="toctext" href>MendelPlots.jl</a><ul class="internal"><li class="toplevel"><a class="toctext" href="#Installation-1">Installation</a></li><li><a class="toctext" href="#Example-dataset-1">Example dataset</a></li><li><a class="toctext" href="#Basic-usage-1">Basic usage</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li><a href>MendelPlots.jl</a></li></ul><a class="edit-page" href="https://github.com/OpenMendel/MendelPlots.jl/blob/master/docs/src/index.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>MendelPlots.jl</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="MendelPlots.jl-1" href="#MendelPlots.jl-1">MendelPlots.jl</a></h1><p>MendelPlots.jl is a Julia package for creating plots for genome-wide association studies (GWAS) results. The package can currently create Manhattan Plots and QQ Plots for GWAS data, specifically catered to the data files created from <a href="https://github.com/OpenMendel"><em>OpenMendel</em></a> software packages <a href="https://github.com/OpenMendel/MendelGWAS.jl">MendelGWAS</a> and <a href="https://github.com/OpenMendel/PolrGWAS.jl">PolrGWAS</a>. The input needed is a dataframe (see <a href="https://github.com/JuliaData/DataFrames.jl">DataFrames.jl</a> or individual features as arrays.</p><p>MendelPlots.jl uses <a href="https://github.com/GiovineItalia/Gadfly.jl">Gadfly</a> as the backend for plotting and saves the plots as .png files. </p><h1><a class="nav-anchor" id="Installation-1" href="#Installation-1">Installation</a></h1><p>This package requires Julia v0.7.0 or later. The package has not yet been registered and must be installed using the repository location. Start julia and use the ] key to switch to the package manager REPL and proceed as follows:</p><pre><code class="language-julia">(v1.0) pkg&gt; add https://github.com/OpenMendel/MendelPlots.jl.git</code></pre><pre><code class="language-julia"># machine information for this tutorial
versioninfo()</code></pre><pre><code class="language-none">Julia Version 1.0.3
Commit 099e826241 (2018-12-18 01:34 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, haswell)</code></pre><p>For use in this tutorial, we will load the following packages:</p><pre><code class="language-julia">using TextParse, DataFrames, MendelPlots</code></pre><pre><code class="language-none">┌ Info: Loading Cairo backend into Compose.jl
└ @ Compose /Users/christophergerman/.julia/packages/Compose/pDFGJ/src/Compose.jl:165
┌ Info: Loading DataFrames support into Gadfly.jl
└ @ Gadfly /Users/christophergerman/.julia/packages/Gadfly/09PWZ/src/mapping.jl:228</code></pre><h2><a class="nav-anchor" id="Example-dataset-1" href="#Example-dataset-1">Example dataset</a></h2><p>Thee <code>data</code> folder of the package in test contains an example simulated data set. In this tutorial, we use relative path <code>../data</code>. In general, the user can locate this folder by the following command:</p><pre><code class="language-julia">import MendelPlots
joinpath(dirname(pathof(MendelPlots)), &quot;../test/data&quot;)</code></pre><pre><code class="language-julia">;ls -l ../test/data</code></pre><pre><code class="language-none">total 832
-rw-r--r--  1 christophergerman  staff  424332 Jan 24 00:05 gwasresults.txt</code></pre><h2><a class="nav-anchor" id="Basic-usage-1" href="#Basic-usage-1">Basic usage</a></h2><p>The following commands can be used to load the test dataset into a dataframe to be used to create plots. </p><pre><code class="language-julia">data, colnames = csvread(&quot;../test/data/gwasresults.txt&quot;, &#39;,&#39;, header_exists = true)
df = DataFrame([i for i in data], Symbol.(colnames))</code></pre><p>&lt;table class=&quot;data-frame&quot;&gt;&lt;thead&gt;&lt;tr&gt;&lt;th&gt;&lt;/th&gt;&lt;th&gt;pval&lt;/th&gt;&lt;th&gt;chr&lt;/th&gt;&lt;th&gt;pos&lt;/th&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;&lt;/th&gt;&lt;th&gt;Float64&lt;/th&gt;&lt;th&gt;Int64&lt;/th&gt;&lt;th&gt;Int64&lt;/th&gt;&lt;/tr&gt;&lt;/thead&gt;&lt;tbody&gt;&lt;p&gt;16,500 rows × 3 columns&lt;/p&gt;&lt;tr&gt;&lt;th&gt;1&lt;/th&gt;&lt;td&gt;0.434119&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;2&lt;/th&gt;&lt;td&gt;0.61908&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;2&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;3&lt;/th&gt;&lt;td&gt;0.909921&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;3&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;4&lt;/th&gt;&lt;td&gt;0.0419107&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;4&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;5&lt;/th&gt;&lt;td&gt;0.344776&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;5&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;6&lt;/th&gt;&lt;td&gt;0.308763&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;6&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;7&lt;/th&gt;&lt;td&gt;0.687524&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;7&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;8&lt;/th&gt;&lt;td&gt;0.0716104&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;8&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;9&lt;/th&gt;&lt;td&gt;0.882899&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;9&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;10&lt;/th&gt;&lt;td&gt;0.762597&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;10&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;11&lt;/th&gt;&lt;td&gt;0.824667&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;11&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;12&lt;/th&gt;&lt;td&gt;0.521365&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;12&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;13&lt;/th&gt;&lt;td&gt;0.95395&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;13&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;14&lt;/th&gt;&lt;td&gt;0.491313&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;14&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;15&lt;/th&gt;&lt;td&gt;0.274445&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;15&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;16&lt;/th&gt;&lt;td&gt;0.0133856&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;16&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;17&lt;/th&gt;&lt;td&gt;0.990837&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;17&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;18&lt;/th&gt;&lt;td&gt;0.782683&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;18&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;19&lt;/th&gt;&lt;td&gt;0.206984&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;19&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;20&lt;/th&gt;&lt;td&gt;0.0443713&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;20&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;21&lt;/th&gt;&lt;td&gt;0.735768&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;21&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;22&lt;/th&gt;&lt;td&gt;0.336091&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;22&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;23&lt;/th&gt;&lt;td&gt;0.810657&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;23&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;24&lt;/th&gt;&lt;td&gt;0.593453&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;24&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;25&lt;/th&gt;&lt;td&gt;0.933379&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;25&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;26&lt;/th&gt;&lt;td&gt;0.867928&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;26&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;27&lt;/th&gt;&lt;td&gt;0.139833&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;27&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;28&lt;/th&gt;&lt;td&gt;0.0694416&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;28&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;29&lt;/th&gt;&lt;td&gt;0.228683&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;29&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;30&lt;/th&gt;&lt;td&gt;0.0627061&lt;/td&gt;&lt;td&gt;1&lt;/td&gt;&lt;td&gt;30&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;&amp;vellip;&lt;/th&gt;&lt;td&gt;&amp;vellip;&lt;/td&gt;&lt;td&gt;&amp;vellip;&lt;/td&gt;&lt;td&gt;&amp;vellip;&lt;/td&gt;&lt;/tr&gt;&lt;/tbody&gt;&lt;/table&gt;</p><p>The basic commands for MendelPlots.jl are </p><pre><code class="language-julia">qq()
manhattan()</code></pre><p>The following command creates a qq plot from the data</p><pre><code class="language-julia">qq(df)</code></pre><p>&lt;img src=&quot;../qqplot.png&quot;&gt;</p><p>For documentation of the <code>qq</code> function, type <code>?qq</code> in Julia REPL.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MendelPlots.qq" href="#MendelPlots.qq"><code>MendelPlots.qq</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">qq(pvalues::AbstractArray)</code></pre><p><strong>Position arguments</strong></p><ul><li><code>pvalues::AbstractArray</code>: pvalues. A one dimensional array containing pvalues to be used in the qqplot.</li></ul><p>qq(df::DataFrame)</p><p><strong>Position arguments</strong></p><ul><li><code>df::DataFrame</code>: DataFrame containing pvalues to be used in the qqplot. Note: The column of the dataframe</li></ul><p>that indicates pvalues must be named pval (df[:pval] must exist)</p><p><strong>Keyword arguments</strong></p><ul><li><code>titles::AbstractString</code>: Title for the plot. Default is &quot;QQ Plot of GWAS p-values&quot;.</li></ul><p>To have blank enter &quot;&quot;. </p><ul><li><p><code>outfile::AbstractString</code>: output name to save for qqplot as a png. Default is &quot;qqplot.png&quot;</p></li><li><p><code>dpi::Union{Float64, Int64}</code>: dots per inch to save the png file. Higher DPI results in larger file with </p></li></ul><p>higher resolution. Default dpi is 350.</p><ul><li><p><code>xlabel::AbstractString</code>: option to replace x-label text</p></li><li><p><code>ylabel::AbstractString</code>: option to replace y-label text</p></li><li><p><code>xmin::Union{Float64, Int64}</code>: Specified minimum x value to represent on the plot</p></li><li><p><code>xmax::Union{Float64, Int64}</code>: Specified maximum x value to represent on the plot</p></li><li><p><code>ymin::Union{Float64, Int64}</code>: Specified minimum y value to represent on the plot</p></li><li><p><code>ymax::Union{Float64, Int64}</code>: Specified maximum y value to represent on the plot</p></li><li><p><code>linecolor::AbstractString</code>: Color of &quot;normal&quot; line. Default color is &quot;red&quot;. </p></li><li><p><code>dotcolor::AbstractString</code>: Color of the dots. Default color is &quot;black&quot;. </p></li><li><p><code>fontsize</code> size of the axis labels. Default is &quot;15pt&quot;. </p></li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/OpenMendel/MendelPlots.jl/blob/245aaf63bfd2caf80b33f144f84e069e908c2e1d/src/gwasplots.jl#L12-L55">source</a></section><p>The following command creates a manhattan plot from the data</p><pre><code class="language-julia">manhattan(df)</code></pre><p>&lt;img src=&quot;../manhattan.png&quot;&gt;</p><p>For documentation of the <code>manhattan</code> function, type <code>?manhattan</code> in Julia REPL.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="MendelPlots.manhattan" href="#MendelPlots.manhattan"><code>MendelPlots.manhattan</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">manhattan(df::DataFrame)</code></pre><p><strong>Position arguments</strong></p><ul><li><code>df::DataFrame</code>: A DataFrame containing information to be used in the Manhattan Plot. </li></ul><p>Note, DataFrame must have the following values saved under the corresponding names.  pvalues:pval, chromosome:chr. Additionally, the DataFrame must be in order of basepairs  going from first to last if there&#39;s no position arguement. Optionally, if there is  basepair information, then the position variable must be named <code>pos</code>. </p><pre><code class="language-none">manhattan(pvalues::AbstractArray, chr::AbstractArray)</code></pre><p><strong>Position arguments</strong></p><ul><li><code>pvalues::AbstractArray</code>: pvalues for the associated GWAS. Must be in the </li></ul><p>order of the basepairs. </p><ul><li><code>chr::AbstractArray</code>: An array of chromosome identifiers for each pvalue. </li></ul><p>Must match order with pvalues. </p><pre><code class="language-none">manhattan(pvalues::AbstractArray, chr::AbstractArray, pos::AbstractArray)</code></pre><p><strong>Position arguments</strong></p><ul><li><code>pvalues::AbstractArray</code>: pvalues for the associated GWAS. </li></ul><p>Must be in the same order of the basepairs and chromosomes. </p><ul><li><code>chr::AbstractArray</code>: An array of chromosome identifiers for each pvalue. </li></ul><p>Must match order with pvalues and positions. </p><ul><li><code>pos::AbstractArray</code>: An array of basepair positions for each pvalue/chromosome. </li></ul><p>Must match order with pvalues and chromosomes. </p><p><strong>Keyword arguments</strong></p><ul><li><code>titles::AbstractString</code>: Title for the plot. Default is &quot;Manhattan Plot&quot;.</li></ul><p>To have blank enter &quot;&quot;. </p><ul><li><code>outfile::AbstractString</code>: output name to save for qqplot as a png.</li></ul><p>Default is &quot;manhattan.png&quot;</p><ul><li><code>dpi::Int64</code>: dots per inch to save the png file. Higher DPI results in </li></ul><p>larger file with higher resolution. Default dpi is 350.</p><ul><li><p><code>xlabel::AbstractString</code>: option to replace x-label text</p></li><li><p><code>ylabel::AbstractString</code>: option to replace y-label text</p></li><li><p><code>ymax::Union{Float64, Int64}</code>: Specified maximum y value to represent </p></li></ul><p>on the plot</p><ul><li><code>signifline::Union{Float64, Int64}</code>: Line to draw significance at. </li></ul><p>Default in Bonferonni corrected p-value for α = 0.05. </p><ul><li><code>linecolor::AbstractString</code>: Color for significance line. Default </li></ul><p>is &#39;deepskyblue1&#39;. </p><ul><li><code>fontsize</code> size of the axis labels. Default is &quot;15pt&quot;. </li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/OpenMendel/MendelPlots.jl/blob/245aaf63bfd2caf80b33f144f84e069e908c2e1d/src/gwasplots.jl#L121-L182">source</a></section><h3><a class="nav-anchor" id="Necessary-Inputs-1" href="#Necessary-Inputs-1">Necessary Inputs</a></h3><p><code>qq</code> expects either an DataFrame object input or an array of pvalues as an input. If a DataFrame is the input, the pvalues must be stored under the name pval.  </p><p><code>manhattan</code> expects either an DataFrame object input or an array of pvalues and an array of chromosome IDs as an input. If a DataFrame is the input, the pvalues must be stored under the name <code>pval</code>.  The chromosomes and pvalues must correspond in order to each other, and must be ordered in according to ascending basepairs. The chromosome variable must be named <code>chr</code>. Optionally, if you have basepair location information in your DataFrame, the position variable must be named <code>pos</code>, but the basepair position isn&#39;t required to create a plot. </p><h3><a class="nav-anchor" id="Additional-Options-1" href="#Additional-Options-1">Additional Options</a></h3><p>There are several other options that the <code>qq</code> and <code>manhattan</code> functions take, refer to the specific documentation for each function via the <code>?</code> command to see the option names. Current options include arguments for qq line color, qq dot color, maximum x and y values, dpi, significance line y-value, significance line color, and title. </p><h3><a class="nav-anchor" id="Output-files-1" href="#Output-files-1">Output files</a></h3><p><code>qq</code> outputs a .png file of the QQ plot. By default, it will be named qqplot.png. Use the keyword argument &#39;outfile&#39; to change the output file name.</p><p><code>manhattan</code> outputs a .png file of the Manhattan plot. By default, it will be named manhattan.png. Use the keyword argument &#39;outfile&#39; to change the output file name.</p><p>You can utilize more options to create a more customized plot. </p><pre><code class="language-julia">qq(df[:pval]; xlabel = &quot;Expected&quot;, ylabel = &quot;Observed&quot;, 
    titles = &quot;&quot;, outfile = &quot;testqq.png&quot;, dotcolor = &quot;gray&quot;, 
    fontsize = 20pt, linecolor = &quot;blue&quot;)</code></pre><p>&lt;img src=&quot;../testqq.png&quot;&gt;</p><footer><hr/></footer></article></body></html>
